File: factory/server/internal/platform/digest/digest.go
// Package digest 对资产正文做 SHA-256 摘要，用于完整性比对，不是内容签名。
package digest

import (
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"

	"github.com/google/uuid"
)

// Sum 计算 32 字节 SHA-256 摘要。
func Sum(content []byte) []byte {
	sum := sha256.Sum256(content)
	out := make([]byte, sha256.Size)
	copy(out, sum[:])
	return out
}

// Match 判断正文与已登记摘要是否一致。
func Match(content, digest []byte) bool {
	got := Sum(content)
	if len(digest) != len(got) {
		return false
	}
	return subtle.ConstantTimeCompare(got, digest) == 1
}

// Member 是闭包摘要的一个成员：身份、修订、已登记摘要和正文。
type Member struct {
	ID       uuid.UUID // 稳定身份
	Revision int64     // 钉死修订
	Digest   []byte    // 该修订已登记摘要
	Content  []byte    // 该修订正文
}

// ClosureSum 按成员顺序拼 id||修订大端||摘要||正文，再 SHA-256。
func ClosureSum(members []Member) []byte {
	h := sha256.New()
	var rev [8]byte
	for _, m := range members {
		id := m.ID
		binary.BigEndian.PutUint64(rev[:], uint64(m.Revision))
		h.Write(id[:])
		h.Write(rev[:])
		h.Write(m.Digest)
		h.Write(m.Content)
	}
	sum := h.Sum(nil)
	out := make([]byte, sha256.Size)
	copy(out, sum)
	return out
}

File: client/app/src/main/kotlin/com/gbndt/shijiaoqi/data/pouch/Digest.kt
package com.gbndt.shijiaoqi.data.pouch

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.UUID

/** SHA-256 与厂端 digest 同形，用于闭包完整性，不是签名。 */
object Digest {
    data class Member(
        val id: UUID,
        val revision: Long,
        val digest: ByteArray,
        val content: ByteArray,
    )

    fun sum(content: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(content)

    fun match(content: ByteArray, digest: ByteArray): Boolean =
        MessageDigest.isEqual(sum(content), digest)

    fun closureSum(members: List<Member>): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        val rev = ByteArray(8)
        for (m in members) {
            md.update(Pouch.uuidBytes(m.id))
            ByteBuffer.wrap(rev).order(ByteOrder.BIG_ENDIAN).putLong(m.revision)
            md.update(rev)
            md.update(m.digest)
            md.update(m.content)
        }
        return md.digest()
    }
}
